File: backend/models/LabTestResult.js
import mongoose from "mongoose";

const testValues = ["pending", "negative", "positive"];
const resultFields = {
	hiv: { type: String, enum: testValues, default: "pending" },
	hbv: { type: String, enum: testValues, default: "pending" },
	hcv: { type: String, enum: testValues, default: "pending" },
	hepatitis: { type: String, enum: testValues, default: "pending" },
	syphilis: { type: String, enum: testValues, default: "pending" },
};

const labTestResultSchema = new mongoose.Schema(
	{
		facility: { type: mongoose.Schema.Types.ObjectId, ref: "Facility", required: true, index: true },
		bloodUnit: { type: mongoose.Schema.Types.ObjectId, ref: "Blood", required: true, unique: true },
		results: { type: new mongoose.Schema(resultFields, { _id: false }), default: () => ({}) },
		status: { type: String, enum: ["draft", "submitted", "approved"], default: "draft" },
		enteredBy: { type: mongoose.Schema.Types.ObjectId, ref: "LabStaff", required: true },
		submittedBy: { type: mongoose.Schema.Types.ObjectId, ref: "LabStaff" },
		submittedAt: Date,
		approvedBy: { type: mongoose.Schema.Types.ObjectId, ref: "LabStaff" },
		approvedAt: Date,
	},
	{ timestamps: true },
);

export default mongoose.model("LabTestResult", labTestResultSchema);

File: backend-go/internal/models/lab_test_result.go
package models

import (
	"go.mongodb.org/mongo-driver/bson/primitive"
	"time"
)

type LabTestResult struct {
	ID        primitive.ObjectID `bson:"_id,omitempty" json:"id,omitempty"`
	Facility  primitive.ObjectID `bson:"facility" json:"facility"`
	BloodUnit primitive.ObjectID `bson:"bloodUnit" json:"bloodUnit"`
	Results   struct {
		HIV       string `bson:"hiv" json:"hiv"` // pending, negative, positive
		HBV       string `bson:"hbv" json:"hbv"`
		HCV       string `bson:"hcv" json:"hcv"`
		Hepatitis string `bson:"hepatitis" json:"hepatitis"`
		Syphilis  string `bson:"syphilis" json:"syphilis"`
	} `bson:"results" json:"results"`
	Status      string              `bson:"status" json:"status"` // draft, submitted, approved
	EnteredBy   primitive.ObjectID  `bson:"enteredBy" json:"enteredBy"`
	SubmittedBy *primitive.ObjectID `bson:"submittedBy,omitempty" json:"submittedBy,omitempty"`
	SubmittedAt *time.Time          `bson:"submittedAt,omitempty" json:"submittedAt,omitempty"`
	ApprovedBy  *primitive.ObjectID `bson:"approvedBy,omitempty" json:"approvedBy,omitempty"`
	ApprovedAt  *time.Time          `bson:"approvedAt,omitempty" json:"approvedAt,omitempty"`
	CreatedAt   time.Time           `bson:"createdAt" json:"createdAt"`
	UpdatedAt   time.Time           `bson:"updatedAt" json:"updatedAt"`
}
